packages/react-dom/src/client/ReactDOMFiberInput.js

```diff
@@ -314,13 +310,39 @@ function synchronizeDefaultValue(
 ) {
   if (
     // Focused number inputs synchronize on blur. See ChangeEventPlugin.js
-    (type !== 'number' || node.ownerDocument.activeElement !== node) &&
-    node.defaultValue !== '' + value
+    type !== 'number' ||
+    node.ownerDocument.activeElement !== node
   ) {
-    if (value != null) {
-      node.defaultValue = '' + value;
-    } else {
-      node.defaultValue = node._wrapperState.initialValue;
-    }
+    assignProperty(
+      node,
+      'defaultValue',
+      value == null ? node._wrapperState.initialValue : value,
+    );
+  }
+}
+
+function assignProperty(
+  node: InputWithWrapperState,
+  property: string,
+  value: *,
+) {
+  // TODO: This should use DOMProperty.shouldSetAttribute, however the current
+  // behavior is such that invalid attributes do not update the current attribute,
+  // instead of removing it.
+  switch (typeof value) {
+    case 'boolean':
+    case 'number':
+    case 'string':
+    case 'object':
+      // Use hasOwnProperty instead of checking for '' to ensure properties
+      // are set for the first time. This is important for defaultValue, which
+      // otherwise does not set an empty value attribute
+      if (!node.hasOwnProperty(property) || node[property] !== '' + value) {
+        // Cast `value` to a string to ensure the value is set correctly. While
+        // browsers typically do this as necessary, jsdom doesn't.
+        // TODO: Is this still true with reasonably modern JSDOM?
+        node[property] = '' + value;
```

Comment:
I'd expect dynamic property access to be slower than just setting a property normally.